crates/ra_hir_ty/src/display.rs

```diff
@@ -6,46 +6,98 @@ use crate::{
     db::HirDatabase, utils::generics, ApplicationTy, CallableDef, FnSig, GenericPredicate,
     Obligation, ProjectionTy, Substs, TraitRef, Ty, TypeCtor,
 };
-use hir_def::{generics::TypeParamProvenance, AdtId, AssocContainerId, Lookup};
+use hir_def::{
+    find_path, generics::TypeParamProvenance, item_scope::ItemInNs, AdtId, AssocContainerId,
+    Lookup, ModuleId,
+};
 use hir_expand::name::Name;
 
-pub struct HirFormatter<'a, 'b> {
+pub struct HirFormatter<'a, R> {
     pub db: &'a dyn HirDatabase,
-    fmt: &'a mut fmt::Formatter<'b>,
+    fmt: &'a mut dyn fmt::Write,
     buf: String,
     curr_size: usize,
     pub(crate) max_size: Option<usize>,
     omit_verbose_types: bool,
+    render_target: R,
 }
 
 pub trait HirDisplay {
-    fn hir_fmt(&self, f: &mut HirFormatter) -> fmt::Result;
+    fn hir_fmt<R>(&self, f: &mut HirFormatter<R>) -> Result<(), R::Error>
+    where
+        R: RenderTarget;
 
-    fn display<'a>(&'a self, db: &'a dyn HirDatabase) -> HirDisplayWrapper<'a, Self>
+    /// Returns a `Display`able type that is human-readable.
+    /// Use this for showing types to the user (e.g. diagnostics)
+    fn display<'a>(
+        &'a self,
+        db: &'a dyn HirDatabase,
+    ) -> HirDisplayWrapper<'a, Self, RenderDiagnostics>
     where
         Self: Sized,
     {
-        HirDisplayWrapper(db, self, None, false)
+        HirDisplayWrapper {
+            db,
+            t: self,
+            max_size: None,
+            omit_verbose_types: false,
+            render_target: RenderDiagnostics,
+        }
     }
 
+    /// Returns a `Display`able type that is human-readable and tries to be succinct.
+    /// Use this for showing types to the user where space is constrained (e.g. doc popups)
     fn display_truncated<'a>(
         &'a self,
         db: &'a dyn HirDatabase,
         max_size: Option<usize>,
-    ) -> HirDisplayWrapper<'a, Self>
+    ) -> HirDisplayWrapper<'a, Self, RenderDiagnostics>
     where
         Self: Sized,
     {
-        HirDisplayWrapper(db, self, max_size, true)
+        HirDisplayWrapper {
+            db,
+            t: self,
+            max_size,
+            omit_verbose_types: true,
+            render_target: RenderDiagnostics,
+        }
+    }
+
+    /// Returns a String representation of `self` that can be inserted into the given module.
+    /// Use this when generating code (e.g. assists)
+    fn render_in_module<'a>(
```

Comment:
Naming: I somehow decided to use the word "render" for the new parts. I'd welcome suggestions for improvement.

Reply:
Alright, after changing to enums, this looked very out of place. I'm calling the new method  `HirDisplay::display_source_code` now